tools/swift-inspect/Sources/swift-inspect/LinuxRemoteProcess.swift

```diff
@@ -0,0 +1,155 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#if os(Linux)
+  import Foundation
+  import SwiftInspectLinux
+  import SwiftRemoteMirror
+
+  internal class LinuxRemoteProcess: RemoteProcess {
+    public typealias ProcessIdentifier = pid_t
+    public typealias ProcessHandle = SwiftInspectLinux.Process
+
+    public private(set) var process: ProcessHandle
+    public private(set) var context: SwiftReflectionContextRef!
+    public private(set) var processIdentifier: ProcessIdentifier
+    public private(set) var processName: String = "<unknown process>"
+
+    let memoryMap: SwiftInspectLinux.MemoryMap
+    let symbolCache: SwiftInspectLinux.SymbolCache
+
+    static var QueryDataLayout: QueryDataLayoutFunction {
+      return { (context, type, _, output) in
+        guard let output = output else { return 0 }
+
+        switch type {
+        case DLQ_GetPointerSize:
+          let size = UInt8(MemoryLayout<UnsafeRawPointer>.stride)
+          output.storeBytes(of: size, toByteOffset: 0, as: UInt8.self)
+          return 1
+
+        case DLQ_GetSizeSize:
+          let size = UInt8(MemoryLayout<UInt>.stride)  // UInt is word-size like size_t
+          output.storeBytes(of: size, toByteOffset: 0, as: UInt8.self)
+          return 1
+
+        case DLQ_GetLeastValidPointerValue:
+          let value: UInt64 = 0x1000
+          output.storeBytes(of: value, toByteOffset: 0, as: UInt64.self)
+          return 1
+
+        default: return 0
+        }
+      }
+    }
+
+    static var Free: FreeFunction {
+      return { (_, bytes, _) in free(UnsafeMutableRawPointer(mutating: bytes)) }
+    }
+
+    static var ReadBytes: ReadBytesFunction {
+      return { (context, address, size, _) in
+        let process: LinuxRemoteProcess = LinuxRemoteProcess.fromOpaque(context!)
+
+        guard
+          let byteArray: [UInt8] = try? process.process.readArray(
+            address: address, upToCount: UInt(size)), let buffer = malloc(byteArray.count)
+        else { return nil }
+
+        byteArray.withUnsafeBytes {
+          buffer.copyMemory(from: $0.baseAddress!, byteCount: byteArray.count)
+        }
+
+        return UnsafeRawPointer(buffer)
+      }
+    }
+
+    static var GetStringLength: GetStringLengthFunction {
+      return { (context, address) in
+        let process: LinuxRemoteProcess = LinuxRemoteProcess.fromOpaque(context!)
+
+        // copy the string from the remote proces to get its length
+        guard let string = try? process.process.readString(address: address),
+          let len = UInt64(exactly: string.count)
```

Comment:
The caller will expect the count in bytes, so this will be short for non-ASCII UTF-8 data, which I think is a realistic possibility if people use non-ASCII in type names or something. `string.utf8.count` should be correct, although I'm mildly worried about normalization coming along and ruining our day. I don't _think_ `String` eagerly normalizes its data though.

Reply:
Ah, ok. Yeah it makes sense the caller wants byte count rather than character count. I will look into it a bit more and use `String.utf8.count` if that looks safe. I could also add a method to just read null-terminated raw bytes.

Reply:
Raw bytes might make more sense, given that this is a C interface with no defined encoding. A `Array<UInt8>` that ends in a `0` would be a decent representation of that.